jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py

```diff
@@ -151,94 +158,106 @@ def to_device(self, index, *args, **kwargs):
+            if self.max_num_training_points is None:
+                self._prefetch_data.extend(list(vecs_iter))
             else:
-                self.logger.info(f'Taking indexed data as training points')
-                while (
-                    self.max_num_training_points
-                    and len(self._prefetch_data) < self.max_num_training_points
-                ):
+                while len(self._prefetch_data) < self.max_num_training_points:
                     try:
                         self._prefetch_data.append(next(vecs_iter))
                     except:
                         break
-                train_data = np.stack(self._prefetch_data)
 
-            if train_data is None:
+            train_data = np.stack(self._prefetch_data)
+            train_data = train_data.astype(np.float32)
+
+            if (
+                self.max_num_training_points
+                and self.max_num_training_points < train_data.shape[0]
+            ):
                 self.logger.warning(
-                    'Loading training data failed. some faiss indexes require previous training.'
+                    f'From train_data with num_points {train_data.shape[0]}, '
+                    f'sample {self.max_num_training_points} points'
                 )
-            else:
-                if (
-                    self.max_num_training_points
-                    and self.max_num_training_points < train_data.shape[0]
-                ):
-                    self.logger.warning(
-                        f'From train_data with num_points {train_data.shape[0]}, '
-                        f'sample {self.max_num_training_points} points'
-                    )
-                    random_indices = np.random.choice(
-                        train_data.shape[0],
-                        size=min(self.max_num_training_points, train_data.shape[0]),
-                        replace=False,
-                    )
-                    train_data = train_data[random_indices, :]
-                train_data = train_data.astype(np.float32)
-                if self.normalize:
-                    faiss.normalize_L2(train_data)
-                self._train(index, train_data)
-
-        self.logger.info(f'Building the faiss {self.index_key} index...')
-        self._build_partial_index(vecs_iter, index)
-        index.nprobe = self.nprobe
-        return index
-
-    def _build_partial_index(self, vecs_iter: Iterable['np.ndarray'], index):
+                random_indices = np.random.choice(
+                    train_data.shape[0],
+                    size=min(self.max_num_training_points, train_data.shape[0]),
+                    replace=False,
+                )
+                train_data = train_data[random_indices, :]
+
+            self.logger.info(f'Training Faiss indexer...')
+
+            if self.normalize:
+                faiss.normalize_L2(train_data)
+            self._train(train_data)
+
+        # TODO: Experimental features
```

Comment:
Do we want to keep this?

Reply:
I think supporting training at initialization would be a good feature for some cases (e.g., small/media data). I would like to keep it.